File: backend/src/app/schemas/chats.py
"""对话/证据相关 Schemas。"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.core.settings import Settings, get_settings
from app.utils.text_sanitization import has_visible_text


class ChatSessionType(str, Enum):
    KB_CHAT = "kb_chat"
    GENERAL_CHAT = "general_chat"


class AgentMode(str, Enum):
    SINGLE_AGENT = "single_agent"
    MULTI_AGENT = "multi_agent"


class MessageRole(str, Enum):
    USER = "user"
    ASSISTANT = "assistant"
    SYSTEM = "system"


class AgentRunStatus(str, Enum):
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CANCELED = "canceled"


class AgentRunType(str, Enum):
    KB_ANSWER = "kb_answer"
    GENERAL_ANSWER = "general_answer"


class EvidenceSourceKind(str, Enum):
    KB = "kb"
    EXTERNAL = "external"


class KbChatConfig(BaseModel):
    """会话级 KB 答案链运行时配置。"""

    model_config = ConfigDict(extra="forbid")

    retrieval_top_k: int = Field(12, ge=1, le=20)
    retrieval_rerank_top_k: int = Field(40, ge=1, le=40)
    retrieval_hybrid_rrf_k: int = Field(60, ge=1, le=200)
    retrieval_parent_max_parents: int = Field(8, ge=1, le=20)
    retrieval_parent_max_children_per_parent: int = Field(3, ge=1, le=10)
    retrieval_multiscale_per_window_top_k: int = Field(40, ge=1, le=200)
    retrieval_multiscale_rrf_k: int = Field(60, ge=1, le=200)
    retrieval_multiscale_max_documents: int = Field(12, ge=1, le=100)
    retrieval_multiscale_max_chunks_per_document: int = Field(2, ge=1, le=20)

    @model_validator(mode="after")
    def validate_constraints(self) -> "KbChatConfig":
        if self.retrieval_rerank_top_k < self.retrieval_top_k:
            raise ValueError("retrieval_rerank_top_k 必须大于等于 retrieval_top_k")
        return self


class KbGraphNode(BaseModel):
    id: str
    label: str
    phase: str | None = None
    order: int | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class KbGraphEdge(BaseModel):
    source: str
    target: str
    conditional: bool = False


class KbGraphSchemaResponse(BaseModel):
    version: str
    hash: str
    nodes: list[KbGraphNode]
    edges: list[KbGraphEdge]


def default_kb_chat_config(*, settings: Settings | None = None) -> KbChatConfig:
    cfg = settings if settings is not None else get_settings()
    return KbChatConfig(
        retrieval_top_k=int(cfg.retrieval_default_top_k),
        retrieval_rerank_top_k=max(
            int(cfg.retrieval_default_top_k),
            int(cfg.retrieval_max_top_k),
        ),
        retrieval_hybrid_rrf_k=int(cfg.retrieval_hybrid_rrf_k),
        retrieval_parent_max_parents=8,
        retrieval_parent_max_children_per_parent=3,
        retrieval_multiscale_per_window_top_k=40,
        retrieval_multiscale_rrf_k=60,
        retrieval_multiscale_max_documents=12,
        retrieval_multiscale_max_chunks_per_document=2,
    )


def resolve_kb_chat_config(
    *,
    raw: KbChatConfig | dict[str, Any] | None,
    settings: Settings | None = None,
) -> KbChatConfig:
    allowed_keys = set(KbChatConfig.model_fields.keys())
    defaults = default_kb_chat_config(settings=settings).model_dump(mode="json")
    if raw is None:
        return KbChatConfig.model_validate(defaults)
    if isinstance(raw, KbChatConfig):
        payload = raw.model_dump(mode="json")
    elif isinstance(raw, dict):
        unknown_keys = set(raw.keys()) - allowed_keys
        if unknown_keys:
            KbChatConfig.model_validate(raw)
        payload = {key: value for key, value in raw.items() if key in allowed_keys}
    else:
        payload = defaults
    merged = {**defaults, **payload}
    return KbChatConfig.model_validate(merged)


# 会话相关
class ChatSessionCreate(BaseModel):
    session_type: ChatSessionType
    selected_kb_ids: list[uuid.UUID] | None = None
    allow_external: bool = False
    mode: AgentMode
    kb_chat_config: KbChatConfig | None = None


class ChatSessionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    session_type: ChatSessionType
    selected_kb_ids: list[uuid.UUID] | None = None
    allow_external: bool
    mode: AgentMode
    kb_chat_config: KbChatConfig | None = None
    created_at: datetime
    updated_at: datetime

    @field_validator("kb_chat_config", mode="before")
    @classmethod
    def _normalize_kb_chat_config(cls, value: Any) -> Any:
        if isinstance(value, dict):
            return resolve_kb_chat_config(raw=value).model_dump(mode="json")
        return value


# 最近对话列表
class ChatSessionRecentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    session_type: ChatSessionType
    title: str | None = None
    updated_at: datetime


WebSearchStatusMode = Literal["healthy", "degraded", "down"]


class WebSearchProviderStatusRead(BaseModel):
    name: Literal["tavily", "searxng", "jina_reader"]
    configured: bool
    verified: bool
    healthy: bool
    mode: WebSearchStatusMode
    latency_ms: int | None = None
    error: str | None = None


class WebSearchStatusRead(BaseModel):
    configured: bool
    verified: bool
    mode: WebSearchStatusMode
    providers: list[WebSearchProviderStatusRead] = Field(default_factory=list)


class ChatRecentListResponse(BaseModel):
    items: list[ChatSessionRecentRead]
    web_search: WebSearchStatusRead


# 消息相关
class ChatMessageCreate(BaseModel):
    content: str = Field(..., min_length=1)
    client_request_id: str | None = Field(default=None, min_length=1, max_length=128)

    @field_validator("content")
    @classmethod
    def _validate_content(cls, value: str) -> str:
        if not has_visible_text(value):
            raise ValueError("content 不能为空")
        return value

    @field_validator("client_request_id", mode="before")
    @classmethod
    def _normalize_client_request_id(cls, value: Any) -> Any:
        if value is None:
            return None
        if not isinstance(value, str):
            raise TypeError("client_request_id 必须是字符串")
        normalized = value.strip()
        if not normalized:
            raise ValueError("client_request_id 不能为空")
        return normalized


class ChatMessageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    role: MessageRole
    content: str
    created_at: datetime


# 证据相关
class EvidenceItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    source_kind: EvidenceSourceKind
    kb_id: uuid.UUID | None = None
    material_id: uuid.UUID | None = None
    chunk_id: uuid.UUID | None = None
    locator: dict | None = None
    excerpt: str
    source_excerpt: str | None = None
    citation_id: str | None = None
    citation_title: str | None = None
    citation_page_hint: str | None = None
    citation_source: str | None = None


# 运行记录相关
class AgentRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    run_type: AgentRunType
    status: AgentRunStatus
    mode: AgentMode
    question: str
    selected_kb_ids: list[uuid.UUID] | None = None
    allow_external: bool
    stage_summaries: dict | None = None
    metrics: dict | None = None
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    error_message: str | None = None


# 两阶段交互：待审批工具调用
class PendingToolCall(BaseModel):
    """待审批的工具调用。"""

    extension_id: str
    extension_name: str | None = None
    tool_name: str
    args: dict[str, Any] = Field(default_factory=dict)
    is_builtin: bool = False


class ToolDecision(BaseModel):
    """单个工具调用的审批决策。"""

    type: Literal["approve", "reject", "edit"]
    message: str | None = None
    edited_action: dict[str, Any] | None = None


class InterruptDecisionBatch(BaseModel):
    """单个 interrupt 的审批决策集合。"""

    interrupt_id: str = Field(..., min_length=1)
    decisions: list[ToolDecision] = Field(default_factory=list, min_length=1)


class ToolApprovalRequest(BaseModel):
    """工具审批请求。"""

    interrupts: list[InterruptDecisionBatch] = Field(default_factory=list, min_length=1)


class PendingInterruptApproval(BaseModel):
    """单个 interrupt 对应的待审批信息。"""

    interrupt_id: str
    message: str | None = None
    pending_tool_calls: list[PendingToolCall] = Field(default_factory=list)


class ClarificationResumeRequest(BaseModel):
    """澄清补充请求。"""

    content: str = Field(..., min_length=1)


# 问答响应（完成）
ClarificationReasonCode = Literal[
    "missing_entity",
    "missing_scope",
    "missing_time",
    "missing_metric",
    "coref_uncertain",
    "mixed",
]


class ClarificationSlot(BaseModel):
    """供用户填写以消除歧义的结构化槽位。"""

    key: str = Field(..., min_length=1, max_length=32)
    label: str = Field(..., min_length=1, max_length=64)
    required: bool = True
    options: list[str] = Field(default_factory=list, max_length=6)


class PendingClarification(BaseModel):
    """供 UI 引导消歧使用的结构化澄清载荷。"""

    question: str = Field(..., min_length=1)
    reason_code: ClarificationReasonCode = "mixed"
    confidence: float = Field(0.0, ge=0.0, le=1.0)
    model_reason: str | None = None
    slots: list[ClarificationSlot] = Field(default_factory=list)
    suggested_answers: list[str] = Field(default_factory=list, max_length=4)


class SemanticCacheMeta(BaseModel):
    hit: bool = False
    score: float | None = Field(default=None, ge=0.0, le=1.0)
    threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    ttl_seconds: int | None = Field(default=None, ge=0)
    entry_id: str | None = Field(default=None, max_length=128)
    schema_version: str | None = Field(default=None, max_length=32)
    hit_type: Literal["strong_hit"] | None = None
    created_at: datetime | str | None = None


class ChatAnswerResponse(BaseModel):
    status: Literal["succeeded"] = "succeeded"
    assistant_message: ChatMessageRead
    evidence: list[EvidenceItem]
    source: Literal["live", "cached"] = "live"
    cache: SemanticCacheMeta | None = None
    stage_summaries: dict | None = None
    metrics: dict | None = None
    run: AgentRunRead


class ChatPendingToolApprovalResponse(BaseModel):
    """两阶段交互的第 1 阶段：返回待审批工具清单。"""

    status: Literal["pending_tool_approval"] = "pending_tool_approval"
    thread_id: str
    pending_interrupts: list[PendingInterruptApproval]
    run: AgentRunRead


class ChatPendingUserClarificationResponse(BaseModel):
    """两阶段交互的第 1 阶段：返回待用户补充澄清信息。"""

    status: Literal["pending_user_clarification"] = "pending_user_clarification"
    thread_id: str
    message: str
    pending_clarification: PendingClarification | None = None
    assistant_message: ChatMessageRead | None = None
    evidence: list[EvidenceItem] = Field(default_factory=list)
    source: Literal["live", "cached"] = "live"
    cache: SemanticCacheMeta | None = None
    stage_summaries: dict | None = None
    metrics: dict | None = None
    run: AgentRunRead

File: frontend/src/services/chats.ts
/**
 * 对话 API 封装
 */

import { apiFetch } from './http';
import { openSseStream } from './sse';
import type { SseEvent } from '../lib/sse';

export type ChatSessionType = 'kb_chat' | 'general_chat';
export type AgentMode = 'single_agent' | 'multi_agent';
export type MessageRole = 'user' | 'assistant' | 'system';
export type AgentRunStatus = 'running' | 'succeeded' | 'failed' | 'canceled';
export type ChatRunStreamStatus = AgentRunStatus | 'waiting_user';
export type TerminalRunStatus = Exclude<ChatRunStreamStatus, 'running'>;
export type EvidenceSourceKind = 'kb' | 'external';
export type ChatMessageResponseStatus =
  | 'succeeded'
  | 'pending_tool_approval'
  | 'pending_user_clarification';

const CHAT_RUN_STREAM_STATUS_VALUES = [
  'running',
  'succeeded',
  'failed',
  'canceled',
  'waiting_user',
] as const;

export interface ChatRunProgress {
  completed: number;
  total: number;
  percent: number;
}

export interface ChatRunStateEvent {
  run_id: string;
  run_status: ChatRunStreamStatus;
  current_step_id: string | null;
  current_step_label: string | null;
  current_step_status: string | null;
  current_node: string | null;
  attempt: number | null;
  message: string | null;
  state_version?: number;
  active_path?: string[];
  last_good_answer?: string | null;
  degrade_reason?: string | null;
  progress: ChatRunProgress;
  ts: string;
}

export interface ChatRunUiEvent {
  event_type: string;
  run_id: string;
  step_id?: string | null;
  status?: string | null;
  node?: string | null;
  message?: string | null;
  candidate_answer?: string | null;
  source_step_id?: string | null;
  degrade_reason?: string | null;
  meta?: Record<string, unknown> | null;
  ts: string;
}

export type ChatTraceExecutionStatus =
  | 'started'
  | 'completed'
  | 'failed'
  | 'waiting_user'
  | 'skipped';

export interface KbChatConfig {
  retrieval_top_k: number;
  retrieval_rerank_top_k: number;
  retrieval_hybrid_rrf_k: number;
  retrieval_parent_max_parents: number;
  retrieval_parent_max_children_per_parent: number;
  retrieval_multiscale_per_window_top_k: number;
  retrieval_multiscale_rrf_k: number;
  retrieval_multiscale_max_documents: number;
  retrieval_multiscale_max_chunks_per_document: number;
}

export interface ChatSessionCreate {
  session_type: ChatSessionType;
  selected_kb_ids?: string[];
  allow_external?: boolean;
  mode: AgentMode;
  kb_chat_config?: KbChatConfig;
}

export interface ChatSession {
  id: string;
  session_type: ChatSessionType;
  selected_kb_ids: string[] | null;
  allow_external: boolean;
  mode: AgentMode;
  kb_chat_config: KbChatConfig | null;
  created_at: string;
  updated_at: string;
}

export interface RecentChatSession {
  id: string;
  session_type: ChatSessionType;
  title: string | null;
  updated_at: string;
}

export interface RecentChatListResponse {
  items: RecentChatSession[];
  web_search_available: boolean;
}

export interface ChatMessage {
  id: string;
  role: MessageRole;
  content: string;
  created_at: string;
}

export interface EvidenceItem {
  source_kind: EvidenceSourceKind;
  kb_id: string | null;
  material_id: string | null;
  chunk_id: string | null;
  locator: Record<string, unknown> | null;
  excerpt: string;
  citation_id?: string | null;
  citation_title?: string | null;
  citation_page_hint?: string | null;
  citation_source?: string | null;
}

export interface AgentRun {
  id: string;
  run_type: string;
  status: AgentRunStatus;
  mode: AgentMode;
  question: string;
  selected_kb_ids: string[] | null;
  allow_external: boolean;
  stage_summaries: Record<string, unknown> | null;
  metrics: Record<string, unknown> | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  error_message: string | null;
}

export interface ChatAnswerResponse {
  status: 'succeeded';
  assistant_message: ChatMessage;
  evidence: EvidenceItem[];
  source?: 'live' | 'cached';
  cache?: SemanticCacheMeta | null;
  stage_summaries?: Record<string, unknown> | null;
  metrics?: Record<string, unknown> | null;
  run: AgentRun;
}

export interface SemanticCacheMeta {
  hit: boolean;
  score?: number | null;
  threshold?: number | null;
  ttl_seconds?: number | null;
}

export interface PendingToolCall {
  extension_id: string;
  extension_name: string | null;
  tool_name: string;
  args: Record<string, unknown>;
  is_builtin: boolean;
}

export interface PendingInterruptApproval {
  interrupt_id: string;
  message: string | null;
  pending_tool_calls: PendingToolCall[];
}

export interface ToolDecision {
  type: 'approve' | 'reject' | 'edit';
  message?: string;
  edited_action?: Record<string, unknown>;
}

export interface InterruptDecisionBatch {
  interrupt_id: string;
  decisions: ToolDecision[];
}

export interface ToolApprovalRequest {
  interrupts: InterruptDecisionBatch[];
}

export interface ChatPendingToolApprovalResponse {
  status: 'pending_tool_approval';
  thread_id: string;
  pending_interrupts: PendingInterruptApproval[];
  run: AgentRun;
}

export type ClarificationReasonCode =
  | 'missing_entity'
  | 'missing_scope'
  | 'missing_time'
  | 'missing_metric'
  | 'coref_uncertain'
  | 'mixed';

export interface ClarificationSlot {
  key: string;
  label: string;
  required: boolean;
  options: string[];
}

export interface PendingClarification {
  question: string;
  reason_code: ClarificationReasonCode;
  confidence: number;
  model_reason?: string | null;
  slots: ClarificationSlot[];
  suggested_answers: string[];
}

export interface ChatPendingUserClarificationResponse {
  status: 'pending_user_clarification';
  thread_id: string;
  message: string;
  pending_clarification?: PendingClarification | null;
  assistant_message?: ChatMessage | null;
  evidence?: EvidenceItem[] | null;
  source?: 'live' | 'cached';
  cache?: SemanticCacheMeta | null;
  stage_summaries?: Record<string, unknown> | null;
  metrics?: Record<string, unknown> | null;
  run: AgentRun;
}

export type ChatMessageResponse =
  | ChatAnswerResponse
  | ChatPendingToolApprovalResponse
  | ChatPendingUserClarificationResponse;

export type ChatStreamEventName =
  | 'meta'
  | 'heartbeat'
  | 'messages'
  | 'updates'
  | 'custom'
  | 'step'
  | 'state'
  | 'ui_event'
  | 'node_io'
  | 'node_trace'
  | 'tool_trace'
  | 'stream_end'
  | 'interrupt'
  | 'final'
  | 'error';

export const KB_CHAT_CUSTOM_EVENT_TYPES = [
  'node_io',
  'answer_review_subcheck',
  'answer_review_fused',
  'guardrail_warning',
  'heartbeat',
] as const;

export type KbChatCustomEventType = (typeof KB_CHAT_CUSTOM_EVENT_TYPES)[number];

export interface NormalizedChatStreamEvent {
  event: ChatStreamEventName;
  version: string;
  payload: Record<string, unknown>;
}

export interface ChatNodeIoEvent {
  execution_id?: string | null;
  display_input_items?: ChatNodeDisplayItem[] | null;
  display_output_items?: ChatNodeDisplayItem[] | null;
  run_id: string;
  node_name: string;
  node_id: string;
  node_path?: string[] | null;
  phase: 'start' | 'end' | 'error';
  attempt?: number | null;
  latency_ms?: number | null;
  input_summary?: Record<string, unknown> | null;
  output_summary?: Record<string, unknown> | null;
  input_snapshot_meta?: Record<string, unknown> | null;
  output_snapshot_meta?: Record<string, unknown> | null;
  input_snapshot?: Record<string, unknown> | null;
  output_snapshot?: Record<string, unknown> | null;
  error_summary?: string | null;
  ts: string;
}

export interface ChatNodeDisplayItem {
  key: string;
  label: string;
  value: string | string[];
}

export interface ChatTraceExecution {
  execution_id: string;
  run_id?: string;
  node_name: string;
  node_label: string;
  status: ChatTraceExecutionStatus;
  started_at: string;
  updated_at: string;
  ended_at?: string | null;
  node_path?: string[] | null;
  attempt?: number | null;
  latency_ms?: number | null;
  input_items?: ChatNodeDisplayItem[] | null;
  output_items?: ChatNodeDisplayItem[] | null;
  error_summary?: string | null;
}

export interface KbGraphNode {
  id: string;
  label: string;
  phase: string | null;
  order: number | null;
  metadata?: Record<string, unknown>;
}

export interface KbGraphEdge {
  source: string;
  target: string;
  conditional: boolean;
}

export interface KbGraphSchema {
  version: string;
  hash: string;
  nodes: readonly KbGraphNode[];
  edges: readonly KbGraphEdge[];
}

export function isChatRunStreamStatus(value: unknown): value is ChatRunStreamStatus {
  return (
    typeof value === 'string' &&
    (CHAT_RUN_STREAM_STATUS_VALUES as readonly string[]).includes(value)
  );
}

export function resolveTerminalRunStatus(
  status: AgentRunStatus | ChatRunStreamStatus | undefined,
  fallback: TerminalRunStatus = 'failed'
): TerminalRunStatus {
  if (!isChatRunStreamStatus(status)) {
    return fallback;
  }
  if (status === 'running') {
    return fallback;
  }
  return status;
}

export function isUnexpectedStreamEnd(params: {
  sawFinalEvent: boolean;
  sawErrorEvent: boolean;
}): boolean {
  return !params.sawFinalEvent && !params.sawErrorEvent;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return null;
  }
  return value as Record<string, unknown>;
}

export function normalizeChatStreamEvent(event: SseEvent): NormalizedChatStreamEvent | null {
  const parsed = (() => {
    try {
      return JSON.parse(event.data) as unknown;
    } catch {
      return null;
    }
  })();
  const record = asRecord(parsed);
  if (!record) {
    return null;
  }

  const explicitType = typeof record.type === 'string' ? record.type : null;
  const normalizedEvent = (explicitType ?? event.event) as ChatStreamEventName;
  const version =
    typeof record.version === 'string'
      ? record.version
      : '2.0';

  return {
    event: normalizedEvent,
    version,
    payload: { ...record },
  };
}

export function toKbGraphSchemaQuery(config: Partial<KbChatConfig>): string {
  const params = new URLSearchParams();
  const orderedKeys: Array<keyof KbChatConfig> = [
    'retrieval_top_k',
    'retrieval_rerank_top_k',
    'retrieval_hybrid_rrf_k',
    'retrieval_parent_max_parents',
    'retrieval_parent_max_children_per_parent',
    'retrieval_multiscale_per_window_top_k',
    'retrieval_multiscale_rrf_k',
    'retrieval_multiscale_max_documents',
    'retrieval_multiscale_max_chunks_per_document',
  ];
  for (const key of orderedKeys) {
    const value = config[key];
    if (typeof value === 'boolean' || typeof value === 'number' || typeof value === 'string') {
      params.set(key, String(value));
    }
  }
  const query = params.toString();
  return query ? `?${query}` : '';
}

/**
 * 创建会话
 */
export async function createChatSession(data: ChatSessionCreate): Promise<ChatSession> {
  return apiFetch<ChatSession>('/api/v1/chats', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

/**
 * 获取会话详情
 */
export async function getChatSession(sessionId: string, signal?: AbortSignal): Promise<ChatSession> {
  return apiFetch<ChatSession>(`/api/v1/chats/${sessionId}`, { signal });
}

/**
 * 删除会话
 */
export async function deleteChatSession(sessionId: string): Promise<void> {
  return apiFetch<void>(`/api/v1/chats/${sessionId}`, {
    method: 'DELETE',
  });
}

/**
 * 获取最近对话
 */
export async function getRecentChats(limit = 20): Promise<RecentChatListResponse> {
  return apiFetch<RecentChatListResponse>(`/api/v1/chats/recent?limit=${limit}`);
}

/**
 * 获取会话消息
 */
export async function getChatMessages(
  sessionId: string,
  signal?: AbortSignal
): Promise<ChatMessage[]> {
  return apiFetch<ChatMessage[]>(`/api/v1/chats/${sessionId}/messages`, { signal });
}

export async function getKbChatGraphSchema(
  config: Partial<KbChatConfig> = {}
): Promise<KbGraphSchema> {
  return apiFetch<KbGraphSchema>(
    `/api/v1/chats/kb-graph-schema${toKbGraphSchemaQuery(config)}`
  );
}

/**
 * 发送消息并获取回答
 */
export async function sendMessage(
  sessionId: string,
  content: string,
  clientRequestId?: string
): Promise<ChatMessageResponse> {
  return apiFetch<ChatMessageResponse>(`/api/v1/chats/${sessionId}/messages`, {
    method: 'POST',
    body: JSON.stringify({
      content,
      client_request_id: clientRequestId,
    }),
  });
}

/**
 * 两阶段交互：提交工具审批并恢复执行
 */
export async function resumeToolApproval(
  sessionId: string,
  runId: string,
  approval: ToolApprovalRequest
): Promise<ChatMessageResponse> {
  return apiFetch<ChatMessageResponse>(`/api/v1/chats/${sessionId}/runs/${runId}/resume`, {
    method: 'POST',
    body: JSON.stringify(approval),
  });
}

/**
 * 发送消息并获取流式回答
 */
export async function streamChatMessage(
  sessionId: string,
  content: string,
  clientRequestId?: string,
  signal?: AbortSignal
): Promise<AsyncIterable<SseEvent>> {
  return openSseStream(
    `/api/v1/chats/${sessionId}/messages/stream`,
    {
      method: 'POST',
      body: JSON.stringify({
        content,
        client_request_id: clientRequestId,
      }),
    },
    signal
  );
}

export async function getPendingGeneralRun(
  sessionId: string,
  signal?: AbortSignal
): Promise<ChatPendingToolApprovalResponse | null> {
  return apiFetch<ChatPendingToolApprovalResponse | null>(
    `/api/v1/chats/${sessionId}/runs/pending-general`,
    { signal }
  );
}

/**
 * 两阶段交互：提交工具审批并恢复执行（流式）
 */
export async function streamResumeToolApproval(
  sessionId: string,
  runId: string,
  approval: ToolApprovalRequest,
  signal?: AbortSignal
): Promise<AsyncIterable<SseEvent>> {
  return openSseStream(
    `/api/v1/chats/${sessionId}/runs/${runId}/resume/stream`,
    {
      method: 'POST',
      body: JSON.stringify(approval),
    },
    signal
  );
}

/**
 * 提交澄清信息并恢复 KB Chat 执行（流式）
 */
export async function streamResumeClarification(
  sessionId: string,
  runId: string,
  content: string,
  signal?: AbortSignal
): Promise<AsyncIterable<SseEvent>> {
  return openSseStream(
    `/api/v1/chats/${sessionId}/runs/${runId}/clarification/stream`,
    {
      method: 'POST',
      body: JSON.stringify({ content }),
    },
    signal
  );
}
